RecordReader for TFRecords

masonk opened the pull request:
I guess I left the work unfinished when I submitted RecordWriter a few years ago. I was hoping some upstream issues would be resolved (e.g., GATs for efficient iteration, and I was expecting the Read trait to change to deal with the uninitialized buffers problem).

I think this is the best that can be done for right now, and it's better to have something than nothing.

masonk commented:
BTW I'm a Googler. Do I still need to sign this with my new email address?

masonk commented on src/io.rs:
```diff
@@ -32,30 +40,147 @@ where
+            Some(len) => len,
+            None => return Ok(None),
+        };
+        let mut vec = vec![0u8; len as usize];
```
I didn't bother with reserving capacity and unsafing to write into uninitialized memory, because I think that performance conscious users are going to use the other method anyway. But let me know what you think.

masonk commented on src/io.rs:
```diff
@@ -32,30 +40,147 @@ where
+    }
+    /// Convert the Reader into an Iterator<Item = Result<Vec<u8>, RecordReadError>, which iterates
+    /// the whole file.
+    pub fn into_iter_owned(self) -> impl Iterator<Item = Result<Vec<u8>, RecordReadError>> {
```
To my knowledge, without GATs, there is no way to write an iterator that owns an internal buffer for reading. Let me know if I'm missing something.

liufuyang replied:
Hmm, I think my knowledge on Rust and TF is not enough yet to give an good answer. Do you perhaps have some other people you can ask for a good review? (Or perhaps we should ask in the community?)

Also I don't have much time recently. Otherwise I will try give a closer look.

masonk commented:
I registered my github account with Google using the corp self-service tool, but that didn't make this bot happy.

masonk commented on src/io.rs:
```diff
@@ -84,14 +209,41 @@ mod tests {
-        let mut expected = vec![0; 0];
-        af.read_to_end(&mut actual).unwrap();
-        ef.read_to_end(&mut expected).unwrap();
+        let _ = std::fs::remove_file(actual_filename);
```
sneak in a line to clean up test files

liufuyang replied:
I am not sure but perhaps we can also create some file in some temp folder? But this seems working as well.

masonk commented:
I squashed these commits, added usage examples in the comments, and changed the email to something that makes cla happy. Please take another look at this PR.

Oh, and I consolidated `InvalidLengthChecksum` and `InvalidContentChecksum` into `InvalidChecksum`, as I couldn't imagine a scenario in which a user would do something different between those cases.

masonk commented on src/io.rs:
```diff
@@ -32,30 +37,181 @@ where
+        let actual_len_crc32 = mask(crc32::checksum_castagnoli(&len_bytes));
+        if expected_len_crc32 != actual_len_crc32 {
+            return Err(RecordReadError::InvalidChecksum);
+        }
```
Done.

masonk commented on src/io.rs:
```diff
@@ -32,30 +37,181 @@ where
+    ///         }
+    ///         Err(e) => { warn!("{:?}", e); break }
+    ///     }
+    /// }
```
I converted my existing test into a doctest and converted these examples into doctests.

masonk commented on src/io.rs:
```diff
@@ -32,30 +37,181 @@ where
+    ///         Err(e) => { warn!("{:?}", e); break }
+    ///     }
+    /// }
+    pub fn read_next(&mut self, buf: &mut [u8]) -> Result<Option<u64>, RecordReadError> {
```
Good catch.

I went with the following approach:

+ pub fn `peek_next_len()`, which tells the length of the next record, and can be used to resize a heap buffer in advance of the `read_next` call.  Rather than wrap into a `BufRead` internally, I added `Seek` bounds, giving control to the caller. In most cases they will be passing `BufRead`s anyway.
+ `read_next()` skips records which are longer than the supplied buffer. So if you don't know required buffer size, peek first

adamcrume commented on src/io.rs:
```diff
@@ -32,30 +39,323 @@ where
+            Some(len) => len,
+            None => return Ok(None),
+        };
+        let mut vec = vec![0u8; len as usize];
```
The buffer shouldn't be allocated until after the length is checksummed, or the program could easily try to allocate a terabyte sized buffer unnecessarily.

adamcrume commented on src/io.rs:
```diff
@@ -32,30 +39,323 @@ where
+                return Err(RecordReadError::CorruptFile);
+            }
+            return Err(RecordReadError::CorruptRecord);
+        }
```
This could be simplified to:

```
if !len_ok {
  Err(RecordReadError::CorruptFile)
} else if !bytes_ok {
  Err(RecordReadError::CorruptRecord)
} else {
  Ok(Some(len as usize))
}
```

masonk replied:
Heh. I think I was a little tired last night.

adamcrume commented on src/io.rs:
```diff
@@ -84,14 +384,95 @@ mod tests {
+                    None => break,
+                },
+                Err(RecordReadError::CorruptFile) | Err(RecordReadError::IoError { .. }) => {
+                    break;
```
Shouldn't this fail the test?

adamcrume commented on src/io.rs:
```diff
@@ -84,14 +384,95 @@ mod tests {
+                assert_eq!(&ary, records[i].as_bytes());
+            } else {
+                break;
+            }
```
Rather than a loop with an if-else that does something different for each index, it would make more sense to unroll it.

adamcrume commented on src/io.rs:
```diff
@@ -32,30 +39,323 @@ where
+                return Err(RecordReadError::CorruptFile);
+            }
+            return Err(RecordReadError::CorruptRecord);
+        }
```
This could be simplified to:

```
if !len_ok {
  Err(RecordReadError::CorruptFile)
} else if !bytes_ok {
  Err(RecordReadError::CorruptRecord)
} else {
  Ok(Some(len as usize))
}
```

adamcrume commented:
Looks good, thanks a lot!

masonk commented:
Thanks for the great reviews

liufuyang commented:
Thank you both very much on this 😄
